From 09a54a862b8f45cff087eb4eabbd283d354afc90 Mon Sep 17 00:00:00 2001
From: Dan Streetman <ddstreet@canonical.com>
Date: Wed, 30 Jun 2021 07:17:22 -0400
Subject: [PATCH 2/3] time: split get_timezone() into main function and
 zone1970.tab function
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1858210
Origin: upstream, https://github.com/systemd/systemd/pull/20066

This allows for adding another function to read from a different timezone
source, which is added in the next commit.
---
 src/basic/time-util.c | 62 ++++++++++++++++++++++++++-----------------
 1 file changed, 38 insertions(+), 24 deletions(-)

--- a/src/basic/time-util.c
+++ b/src/basic/time-util.c
@@ -1202,7 +1202,7 @@ bool ntp_synced(void) {
         return true;
 }
 
-int get_timezones(char ***ret) {
+static int get_timezones_from_zone1970_tab(char ***ret) {
         _cleanup_fclose_ FILE *f = NULL;
         _cleanup_strv_free_ char **zones = NULL;
         int r;
@@ -1210,35 +1210,49 @@ int get_timezones(char ***ret) {
         assert(ret);
 
         f = fopen("/usr/share/zoneinfo/zone1970.tab", "re");
-        if (f) {
-                for (;;) {
-                        _cleanup_free_ char *line = NULL, *cc = NULL, *co = NULL, *tz = NULL;
-
-                        r = read_line(f, LONG_LINE_MAX, &line);
-                        if (r < 0)
-                                return r;
-                        if (r == 0)
-                                break;
-
-                        const char *p = line;
-
-                        /* Line format is:
-                         * 'country codes' 'coordinates' 'timezone' 'comments' */
-                        r = extract_many_words(&p, NULL, 0, &cc, &co, &tz, NULL);
-                        if (r < 0)
-                                continue;
-
-                        /* Lines that start with # are comments. */
-                        if (*cc == '#')
-                                continue;
-
-                        r = strv_extend(&zones, tz);
-                        if (r < 0)
-                                return r;
-                }
-        } else if (errno != ENOENT)
+        if (!f)
                 return -errno;
 
+        for (;;) {
+                _cleanup_free_ char *line = NULL, *cc = NULL, *co = NULL, *tz = NULL;
+
+                r = read_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                const char *p = line;
+
+                /* Line format is:
+                 * 'country codes' 'coordinates' 'timezone' 'comments' */
+                r = extract_many_words(&p, NULL, 0, &cc, &co, &tz, NULL);
+                if (r < 0)
+                        continue;
+
+                /* Lines that start with # are comments. */
+                if (*cc == '#')
+                        continue;
+
+                r = strv_extend(&zones, tz);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(zones);
+        return 0;
+}
+
+int get_timezones(char ***ret) {
+        _cleanup_strv_free_ char **zones = NULL;
+        int r;
+
+        assert(ret);
+
+        r = get_timezones_from_zone1970_tab(&zones);
+        if (r < 0 && r != -ENOENT)
+                return r;
+
         /* Always include UTC */
         r = strv_extend(&zones, "UTC");
         if (r < 0)
